File: Data Structures & Algorithms/maximum-product-subarray/submission-1.py
class Solution:
    # Kadane's algorithm, O(n) time, O(1) space
    def maxProduct(self, nums: List[int]) -> int:
        max_product = nums[0]
        curr_max, curr_min = 1, 1

        for num in nums:
            temp = curr_max * num # we want to keep track of the previous curr_max
            # because we modify curr_max here, and we still want to use the previous one to compute curr_min
            curr_max = max(num, num * curr_max, num * curr_min)
            curr_min = min(num, temp, num * curr_min)
            max_product = max(curr_max, max_product)

        return max_product

File: Data Structures & Algorithms/number-of-good-pairs/submission-1.cpp
class Solution {
public:
    // frequency counter solution, O(n) time and space
    int numIdenticalPairs(vector<int>& nums) {
        unordered_map<int, int> counter;
        int numPairs = 0;
        for (int num : nums) {
            numPairs += counter[num];
            counter[num]++;
        }
        return numPairs;
    }
};